### PR TITLE
CLI: Add split and merge subcommand

### PR DESCRIPTION
**Pull request by wkalt**
Adds subcommands "split" and "merge". Split splits an mcap file into one
output file per channel. Merge merges a set of mcap files into a single
output. For instance,

    mcap split demo.mcap -o output-dir
    mcap merge output-dir/* > output.mcap

Output of merge can be piped to other tools, e.g

    mcap merge output-dir/* | mcap cat --json

### CODE REVIEW

**Comment by wkalt** [comment]
I added split here mostly to easily test the functionality interactively, but it also occurred to me we could wrap split as a CLI --separate-outputs flag on the filter subcommand requested here: https://github.com/foxglove/mcap/issues/351

This would more naturally allow extension to do stuff like split out only certain topics, or all but certain. Once we add that I think split will start looking pretty similar to filter.

File: go/cli/mcap/cmd/merge.go [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+			return fmt.Errorf("failed to add initial schema for input %d: %w", inputID, err)
+		}
+		newChannelID, err := m.addChannel(writer, inputID, channel)
+		if err != nil {
+			return fmt.Errorf("failed to add initial channel for input %d: %w", inputID, err)
+		}
+		iterators[inputID] = iterator
+
+		// use the mapped output channel ID when pushing to the heap
+		message.ChannelID = newChannelID
```
Comment (wkalt) [inline go/cli/mcap/cmd/merge.go:179]: that can be assigned on 172

File: go/cli/mcap/cmd/merge.go [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+		// the message to be written. This is numbered with the correct channel
+		// ID for the output, and schemas + channels for it have already been
+		// written, so it can be written straight to the output.
+		msg := heap.Pop(pq).(utils.TaggedMessage)
+		err = writer.WriteMessage(msg.Message)
+		if err != nil {
+			return fmt.Errorf("failed to write message: %w", err)
+		}
+
+		// the next message off the iterator, to replace the one just popped
```
Comment (wkalt) [inline go/cli/mcap/cmd/merge.go:196]: "Pull the next..."

**Comment by amacneil** [comment]
Seems like both commands should support the `-o` flag (if passed to `merge`, it would write to the specified file rather than stdout)

File: go/cli/mcap/utils/priority_queue.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+type PriorityQueue []TaggedMessage
+
+func (pq PriorityQueue) Len() int {
+	return len(pq)
+}
+
+func (pq PriorityQueue) Less(i, j int) bool {
+	return pq[i].Message.LogTime < pq[j].Message.LogTime
```
Comment (wkalt) [inline go/cli/mcap/utils/priority_queue.go:16]: this should do a secondary comparison on import ID so ties are broken deterministically

**Comment by wkalt** [comment]
the formatting of the `split` output file names should be noted somewhere.. it's `<topic>_channelID.mcap`, where the topic has slashes converted to double underscores. Some other ideas would be,
* group files by schema instead of channel (I think bad, because the schema name will not be what users want to look for)
* group all channels for the same topic name into the same file, and name it `<topic>.mcap` with the same transformation
* prefix the topic names with the channel ID instead of suffixing them. Maybe nicer filenames, but also don't sort on topic name.

File: go/cli/mcap/cmd/merge.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+	// for each input reader, initialize an mcap reader and read the first
+	// message off. Insert the schema and channel into the output with
+	// renumbered IDs, and load the message (with renumbered IDs) into the
+	// priority queue.
+	for inputID, inputReader := range inputs {
+		reader, err := mcap.NewReader(inputReader)
+		if err != nil {
+			return fmt.Errorf("failed to create input reader for input %d: %w", inputID, err)
+		}
+		iterator, err := reader.Messages(0, math.MaxInt64, nil, false)
```
Comment (wkalt) [inline go/cli/mcap/cmd/merge.go:156]: instantiation of the reader/iterator can probably be pulled into a method

**Comment by amacneil** [comment]
What if you let the user specify the file naming. e.g.:

```sh
mcap split demo.mcap -o "outputdir/{channel_topic}_{channel_id}.mcap"
```

**Comment by wkalt** [comment]
closing - work moved/merged in #418

**Pull request closed without merging**